Realm/ObjectStore/handover.hpp

```diff
@@ -0,0 +1,120 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2015 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef handover_hpp
+#define handover_hpp
+
+#include <realm/row.hpp>
+#include <realm/query.hpp>
+#include <realm/group_shared.hpp>
+#include <realm/link_view.hpp>
+#include <realm/table_view.hpp>
+
+namespace realm {
+    class HandoverPackage;
+    class SharedGroup;
+
+    class AnyHandover;
+
+    // Type-erased wrapper for any type which must be exported to be handed between threads
+    class AnyThreadConfined {
+    public:
+        enum class Type {
+            Row,
+            Query,
+            TableRef,
+            TableView,
+            LinkViewRef,
+        };
+
+    private:
+        Type m_type;
+        union {
+            Row m_row;
+            Query m_query;
+            TableRef m_table_ref;
+            TableView m_table_view;
+            LinkViewRef m_link_view_ref;
+        };
+
+    public:
+        // Constructors
+        AnyThreadConfined(Row row)                   : m_type(Type::Row),         m_row(row)                     { }
+        AnyThreadConfined(Query query)               : m_type(Type::Query),       m_query(query)                 { }
+        AnyThreadConfined(TableRef table_ref)        : m_type(Type::TableRef),    m_table_ref(table_ref)         { }
+        AnyThreadConfined(TableView table_view)      : m_type(Type::TableView),   m_table_view(table_view)       { }
+        AnyThreadConfined(LinkViewRef link_view_ref) : m_type(Type::LinkViewRef), m_link_view_ref(link_view_ref) { }
+
+        AnyThreadConfined(const AnyThreadConfined& thread_confined);
+        AnyThreadConfined(AnyThreadConfined&& thread_confined);
+        ~AnyThreadConfined();
+
+        inline Type type() const { return m_type; }
```

Comment:
`inline` here is redundant (member functions defined inline are implicitly `inline`, as otherwise including the header in multiple cpp files would always be a linker error).